Refactor demo server (#271)
trento-project/wanda: guides/demo.md

## How to run Wanda in Demo mode?

Instead of starting the local developer environment we need to start wanda in demo mode we can use this command

```
DATABASE_URL=ecto://postgres:postgres@localhost:5434/wanda_dev AMQP_URL=amqp://wanda:wanda@localhost:5674 SECRET_KEY_BASE=Tbp26GilFTZOXafb7FNVqt4dFQdeb7FAJ++am7ItYx2sMzYaPSB9SwUczdJu6AhQ CORS_ORIGIN=http://localhost:4000 ACCESS_TOKEN_ENC_SECRET=s2ZdE+3+ke1USHEJ5O45KT364KiXPYaB9cJPdH3p60t8yT0nkLexLBNw8TFSzC7k  PORT=4001  MIX_ENV=demo iex -S mix phx.server
```

Explanation:

- DATABASE_URL: The URL to connect to the Postgres database running in the Docker container.
- AMQP_URL: The URL for RabbitMQ, used for message exchange.
- SECRET_KEY_BASE: The secret key for Wanda, keep it secure.
- CORS_ORIGIN: The origin URL from where API requests are allowed (pointing to web).
- ACCESS_TOKEN_ENC_SECRET: Secret key for encrypting access tokens.
- PORT: The port on which Wanda will run (4001 in this case).
- MIX_ENV: The environment in which Wanda will run (set to "demo" for the demo environment).

> ## Modify the demo facts configuration

Reviewer: I would probably focus on the structure of the yaml and its meaning first, cause then it would be easier to grasp how to add things

```yaml
targets:
  <target_reference>: <target_id>

facts:
  <check_id>:
    <fact_name>:
      <target_reference>: <Fact Value>
    <another_fact_name>:
      <another_target_reference>: <Fact Value>
```

where the `targets` section allows to define handles for targets' uuids so that we do not have to repeat uuids all around and keep the yaml less noisy.

while the `facts` section allows to define which `<Fact Value>` a target (referenced by `<target_reference>`) should return for a specific fact (`<fact_name>`) on a given check (`<check_id>`).

Then we can say that we can add as many targets as needed and as many checks/facts as needed.

- PORT: The port on which Wanda will run (4001 in this case).
- MIX_ENV: The environment in which Wanda will run (set to "demo" for the demo environment).

## Modify demo facts configuration

In the Wanda demo environment, configure fake facts by editing the [fake_facts configuration](/priv/demo/fake_facts.yaml). Define or modify custom facts for different targets and checks.

### YAML Structure

The Configuration is saved in [fake_facts.yaml](/priv/demo/fake_facts.yaml) file, which follows a specific structure with two main sections: targets and facts.

Example:

```yaml
targets:
  target1: 0a055c90-4cb6-54ce-ac9c-ae3fedaf40d4
  target2: 13e8c25c-3180-5a9a-95c8-51ec38e50cfc
facts:
  check_1:
    fact_name1:
      target1: 2
      target2: 3
```

#### Targets

The targets section allows defining handles for targets' UUIDs, providing a reference that can be used through the configuration.

Format:

```yaml
targets:
  <target_reference>: <target_id>
```

<target_reference>: A user-defined handle or reference for a target.

<target_id>: The UUID or identifier of the target.

#### Add a new target to the demo configuration

> Open the [fake_facts](priv/demo/fake_facts.yaml) configuration and add a new target entry to the targets section, following this format:

Reviewer: absolute url

```
targetX: <new_target_id>
```

Replace targetX with an unique target identifier (e.g target3) and <new_target_id>

Add the new target entry:
```yaml
targets:
  target1: 0a055c90-4cb6-54ce-ac9c-ae3fedaf40d4
  target2: 13e8c25c-3180-5a9a-95c8-51ec38e50cfc
  target3: <new_target_id>
```

### Facts Section:

The facts section allows specifying the facts associated with specific targets and checks. Each fact corresponds to a piece of information that a target can return during the evaluation of a check.

The format for the facts section is as follows:

```yaml
facts:
  <check_id>:
    <fact_name>:
      <target_reference>: <fact value>
      <target_reference2>: <fact value>
```

> #### Add a new Fact to the demo configuration

Reviewer: As before we could simplify by saying that:
- there can be as many `<check_id>` as needed
- each `<check_id>` there can be as many `<fact_name>` configured as needed
- each `<fact_name>` can be instrumented to have a specific `<fact value>` on a `<target_reference>`

Or something along these lines.
Sounds more straightforward.

> Open the [fake_facts](priv/demo/fake_facts.yaml) configuration and add a new fact to the facts section, following this format:

Reviewer: absolute url

> Open the [fake_facts](/priv/demo/fake_facts.yaml) configuration and add a new fact to the facts section, following this format:

Reviewer: in any case absolute URLs

```yaml
facts:
  <check_id>:
    <fact_name>:
      target1: <fact value>
      target2: <fact value>
      target3: <fact value>
```

Example:

```yaml
targets:
  target1: 0a055c90-4cb6-54ce-ac9c-ae3fedaf40d4
  target2: 13e8c25c-3180-5a9a-95c8-51ec38e50cfc
  target3: <new_target_id>

facts:
.......existing facts..........
  new_check_id:
    new_fact_name:
      target1: <fact value>
      target2: <fact value>
      target3: <fact value>
```

## FAQ

> ### How to apply the configuration changes?

Reviewer: Not really necessary imho.

Save the [fake_facts](/priv/demo/fake_facts.yaml) configuration file and re-run executions from the [web](https://github.com/trento-project/web) ui

### What happens with unmapped targets, facts or checks?

Wanda will still evaluate and return a default fallback fact value "some fact value" for a check's execution. This means when a new check is added to wanda, changing the the configuration is optional.

> ### How to know what's the correct return value of a fact?

Reviewer: We can shrink this as well by providing just a reference to the gatherers and the catalog.

See [link to gatherers] and [link to catalog] for extra information about what checks expect.

Author: Done!

Example:

```yaml
targets:
  target1: 0a055c90-4cb6-54ce-ac9c-ae3fedaf40d4
  target2: 13e8c25c-3180-5a9a-95c8-51ec38e50cfc
facts:
  check_1:
    fact_name1:
      target1: 2
      target2: 3
```

#### Targets

The targets section allows defining handles for targets' UUIDs, providing a reference that can be used through the configuration.

Format:

```yaml
targets:
  <target_reference>: <target_id>
```

<target_reference>: A user-defined handle or reference for a target.

<target_id>: The UUID or identifier of the target.

#### Add a new target to the demo configuration

Open the [fake_facts](priv/demo/fake_facts.yaml) configuration and add a new target entry to the targets section, following this format:

```
targetX: <new_target_id>
```

Replace targetX with an unique target identifier (e.g target3) and <new_target_id>

Add new target with an identifier (e.g target3) and a new target id:

```
targets:
  target1: 0a055c90-4cb6-54ce-ac9c-ae3fedaf40d4
  target2: 13e8c25c-3180-5a9a-95c8-51ec38e50cfc
  target3: <new_target_id>
```

### Facts Section:

> The facts section allows specifying the facts associated with specific targets and checks. Each fact corresponds to a piece of information that a target can return during the evaluation of a check.

Reviewer: > Each fact corresponds to a piece of information that a target can return during the evaluation of a check.
>
This is not totally correct. Facts are what targets return from facts gathering, not during the evaluation.
Evaluation is something that happens after fact gathering, which uses gathered facts in evaluation.

```suggestion
The facts section allows specifying what fact value a target should return for a specific check.
```
